feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/clustering/spi/main/module.xml

```diff
@@ -33,14 +33,12 @@
 
     <dependencies>
         <module name="org.jboss.as.clustering.common"/>
-        <module name="org.jboss.as.controller"/>
-        <module name="org.jboss.modules"/>
-        <module name="org.jboss.msc"/>
         <module name="org.wildfly.clustering.api"/>
-        <module name="org.wildfly.clustering.ejb.infinispan" services="import"/>
-        <module name="org.wildfly.clustering.server" services="import"/>
+        <!-- Enable dynamic installation of clustering services for EJB affinity support -->
+        <module name="org.wildfly.clustering.ejb.infinispan" services="import" optional="true"/>
+        <!-- Enable dynamic installation of clustering services per channel or cache -->
+        <module name="org.wildfly.clustering.server" services="import" optional="true"/>
```

Comment:
In my testing work I found I needed to ensure Galleon provisioned this if the JGroups subsystem was used, i.e. via

https://github.com/wildfly/wildfly/pull/12508/commits/a23ae9cf642eced338d5149353c422d38cfea2e7#diff-4f073585cd66264ee958d5a02ee1d60bR272

I'm not sure if that will still be necessary given all the other changes here but I suspect it will. I'm rebasing my branch onto this one and will see what happens.

Reply:
Yep, testsuite/layers LayersTests fails trying to boot a server provisioned with 'jpa-distributed'.

```
22:05:31,821 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-1) MSC000001: Failed to start service org.wildfly.clustering.jgroups.channel-module.ee: org.jboss.msc.service.StartException in service org.wildfly.clustering.jgroups.channel-module.ee: java.lang.IllegalArgumentException: org.jboss.modules.ModuleNotFoundException: org.wildfly.clustering.server
	at org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:70)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1738)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1700)
	at org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1558)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: org.jboss.modules.ModuleNotFoundException: org.wildfly.clustering.server
	at org.jboss.as.clustering.controller.ModuleServiceConfigurator.get(ModuleServiceConfigurator.java:80)
	at org.jboss.as.clustering.controller.ModuleServiceConfigurator.get(ModuleServiceConfigurator.java:48)
	at org.wildfly.clustering.service.FunctionalService.start(FunctionalService.java:67)
	... 8 more
Caused by: org.jboss.modules.ModuleNotFoundException: org.wildfly.clustering.server
	at org.jboss.modules.ModuleLoader.loadModule(ModuleLoader.java:297)
	at org.jboss.as.clustering.controller.ModuleServiceConfigurator.get(ModuleServiceConfigurator.java:78)
	... 10 more

22:05:31,934 ERROR [org.jboss.as.controller.management-operation] (Controller Boot Thread) WFLYCTL0013: Operation ("add") failed - address: ([
    ("subsystem" => "jgroups"),
    ("channel" => "ee")
]) - failure description: {"WFLYCTL0080: Failed services" => {"org.wildfly.clustering.jgroups.channel-module.ee" => "java.lang.IllegalArgumentException: org.jboss.modules.ModuleNotFoundException: org.wildfly.clustering.server
    Caused by: java.lang.IllegalArgumentException: org.jboss.modules.ModuleNotFoundException: org.wildfly.clustering.server
    Caused by: org.jboss.modules.ModuleNotFoundException: org.wildfly.clustering.server"}}
```

Reply:
That commit I linked above is admittedly a convoluted way to do that. The org.jboss.as.clustering.jgroups module.xml could just have the dep, with a comment about why it's needed

Reply:
The attempt to load this module is loaded based on the default value of this attribute:
https://github.com/wildfly/wildfly/blob/master/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/ChannelResourceDefinition.java#L112

Reply:
@pferraro Yep. That's been there for at least 2 years, so to remove it would require some sort of compatibility handling.  If we don't do that we need to be sure the module gets provisioned.

BTW, my PR layered on top of this got good results -- https://github.com/wildfly/wildfly/pull/12508

Both in the standard CI jobs and in the custom jobs that test against Galleon slimmed servers:

https://ci.wildfly.org/viewLog.html?buildId=167028&

https://ci.wildfly.org/viewLog.html?buildId=167029&

But that had a commit to deal with this issue. I think that problem would be better addressed in this PR.  Other than that this LGTM.

Reply:
> That's been there for at least 2 years, so to remove it would require some sort of compatibility handling. If we don't do that we need to be sure the module gets provisioned.

We wouldn't want to remove it - as this attribute configures the module from which channel-based services configure their marshaller.

What I meant to say is that we don't want to hardcode the module to be loaded in the ResourceDefinition constructor - as this value is dynamic.

Reply:
@pferraro There's no facility for telling Galleon a module is needed based on the value of an attribute. If a user wants to refer to a custom module they'll need to ensure Galleon provisions it by explicitly declaring the package in their provisioning.xml or making it a default package in config or layer they provide. But in this default case we don't want them to have to do that so I think we need to hard code it.

@jfdenise @ehsavoie I don't know if there's a possible future enhancement here; i.e. AttributeDefinition.getArbitraryDescriptors() can look for a "module-name" arbitrary descriptor and if found put something in the feature spec which galleon somehow uses to require a package that matches the configured value.  That couldn't work though if the attribute allowed an expression, which this one does.

Reply:
@bstansberry OK.  So that means it needs to be non-optional in module.xml.

Reply:
Addressed by https://github.com/wildfly/wildfly/pull/12614/commits/19401d7fd1392c123d850e2e4e114db648cde772